src/Microsoft.AspNet.JsonPatch/Adapters/SimpleObjectAdapter.cs

```diff
@@ -644,70 +575,96 @@ public void Copy(Operation<T> operation, T objectToApplyTo)
+                                operation.from),
+                            objectToApplyTo);
                     }
 
                     valueAtFromLocation = array[positionAsInteger];
-
                 }
                 else
                 {
                     throw new JsonPatchException<T>(operation,
-                       string.Format("Patch failed: provided from path is invalid for array property type at location from: {0}: expected array",
-                       operation.from),
+                       string.Format("Patch failed: provided from path is invalid for array property type at " +
+                            "location from: {0}: expected array",
+                            operation.from),
                        objectToApplyTo);
                 }
             }
             else
             {
                 // no list, just get the value
-
                 // set the new value
-
-                valueAtFromLocation = PropertyHelpers.GetValue(fromProperty, objectToApplyTo, actualFromProperty);
-
+                valueAtFromLocation = patchProperty.Property.ValueProvider.GetValue(patchProperty.Parent);
             }
 
             // add operation to target location with that value.
-
             Add(operation.path, valueAtFromLocation, objectToApplyTo, operation);
+        }
 
+        private void CheckIfPropertyExists(
+            JsonPatchProperty patchProperty,
+            T objectToApplyTo,
+            Operation<T> operation,
+            string propertyPath)
+        {
+            if (patchProperty == null)
+            {
+                throw new JsonPatchException<T>(operation,
+                    string.Format("Patch failed: property at location {0} does not exist", propertyPath),
+                    objectToApplyTo);
+            }
+            if (patchProperty.Property.Ignored)
+            {
+                throw new JsonPatchException<T>(operation,
+                    string.Format("Patch failed: cannot update property at location {0}", propertyPath),
+                    objectToApplyTo);
+            }
+        }
+
+        private bool IsNonStringArray(JsonPatchProperty patchProperty)
+        {
+            return !(patchProperty.Property.PropertyType == typeof(string))
+                    && typeof(IList).GetTypeInfo().IsAssignableFrom(
+                        patchProperty.Property.PropertyType.GetTypeInfo());
+        }
+
+        private void CheckIfPropertyCanBeSet(ConversionResult result,
+            T objectToApplyTo,
+            Operation<T> operation,
+            string path)
+        {
+            if (!result.CanBeConverted)
+            {
+                throw new JsonPatchException<T>(operation,
+                            string.Format("Patch failed: provided value is invalid for property type at " +
+                                "location path: {0}",
+                                path),
+                            objectToApplyTo);
+            }
```

Comment:
Clean up formatting on all of these private methods, parameters on their own lines, use local variables where appropriate, fix indenting.